src/lib/updating-element.ts

```diff
@@ -0,0 +1,418 @@
+      if (attrValue !== undefined) {
+        // TODO(sorvell): doing this here means we need to allow round tripping
+        // through property setting to not interfere with the reaction stack.
+        // However, we could move this to update (as in Polymer) to avoid this
+        // with a flag, but this is undesirable since we want to keep update
+        // unimplemented here.
+        this.setAttribute(name, attrValue);
+      }
+      this.invalidate();
+    }
+  }
+
+  /**
+   * Call to request the element to asynchronously update regardless
+   * of whether or not any property changes are pending. This method is
+   * automatically called when any registered property changes.
+   */
+  protected async invalidate() {
+    if (this._isEnabled) {
+      if (this._isUpdating) {
+        console.warn('Requested an update while updating. This is not supported.');
+      } else if (!this._isValidating) {
+        this._isValidating = true;
+        await Promise.resolve();
+        this._validate();
+      }
+    }
+  }
+
+  /**
+   * Validates the element. If `shouldUpdate` returns true then `update` is
+   * called. When the update is complete `changeProps` and `prevProps` are
+   * updated and any `updateComplete` promise is resolved.
+   */
+  private _validate() {
+    this._isValidating = false;
+    if (!this.shouldUpdate()) {
+      return;
+    }
+    this._isUpdating = true;
+    this.update();
+    this._hasUpdated = true;
+    this.changedProps = {};
+    Object.assign(this.prevProps, this.props);
+    this._isUpdating = false;
+    if (this._updateCompleteResolver) {
+      this._updateCompleteResolver(true);
+    }
+  }
+
+  /**
+   * Implement to control if updating should occur when property values
+   * change or `invalidate` is called. By default, this method always
+   * returns true, but this can be customized as an optimization to avoid
+   * rendering work when changes occur which should not be rendered.
+   */
+  protected shouldUpdate(): boolean {
+    return true;
+  }
+
+  /**
+   * Implement to update the element. Typically renders and keeps updated DOM
+   * in the element's root.
+   */
+  protected update() {
+  }
+
+ /**
+   * Returns a promise which resolves after the element next updates.
+   * The promise resolves to `true` if the element updated and `false` if the
+   * element did not.
+   * This is useful when users (e.g. tests) need to react to the updated state
+   * of the element after a change is made.
+   * This can also be useful in event handlers if it is desireable to wait
+   * to send an event until after updating.
+   */
+  get updateComplete() {
+    if (!this._updatePromise) {
+      this._updatePromise = new Promise((resolve) => {
+        this._updateCompleteResolver = (value: boolean) => {
+          this._updateCompleteResolver = this._updatePromise = null;
+          resolve(value);
+        };
+      });
+      if (!this._isUpdating) {
```

Comment:
This behavior should be documented, but I also wondering if it's what we want. Should this not wait till after the next update?